# Homework 3  - Tryout pandas and numpy

Copy this notebook. Rename it as: YOURNAME-Homework-panda-numpy-Sept-2017

with your name replacing yourname and  replaced with the date you submit or to the last part.

Do the homeworks problems below.

Upload your completed jupyter notebook to your github site and send me the url via the elearning site as your homework submission. Please note that you need to do your own 3.4, and 3.5 if you share data and analysis from others then you cannot get more than a 3.  

## Homework 3.1

### 3.1.a
Load the data from: http://opendata.dc.gov/datasets that I have include in this github   
into a dataframe.  ( The file has been is available in directory ./data/ccp_current_csv.csv )

In [2]:
import pandas as pd

df = pd.read_csv("ccp_current_csv.csv")

df.head()

STREET QUADRANT  FROMINTERSECTION              TOINTERSECTION  \
0           13th St       NW        Kennedy St               Longfellow St   
1              S St       NW  North Capitol St                      3rd St   
2  Rhode Island Ave       NE  Brentwood Rd (W)            Brentwood Rd (M)   
3           Mill Rd       NW           27th St                  Cul De Sac   
4  Rhode Island Ave       NE  Brentwood Rd (M)  14th St  Evarts St  Montan   

   NOOFBLOCKS                 WORKDESCRIPTION  YEARBUDGETED  PERCENTCOMPLETED  \
0           1  Pavement Resurface/restoration          2004               100   
1           3                   Regular Cover          1999               100   
2           1  Pavement Resurface/restoration          2004               100   
3           1  Pavement Resurface/restoration          2004               100   
4           2  Pavement Resurface/restoration          2005               100   

      STATUS  MILES      ...             YCOORD   LATITUDE  LONGITUDE  \
0  Completed   0.07      ...       143205.11981  38.956747 -77.029694   
1  Completed   0.33      ...       138468.90256  38.914085 -77.012411   
2  Completed   0.09      ...       139574.10659  38.924041 -76.987221   
3  Completed   0.07      ...       138132.90198  38.911046 -77.054077   
4  Completed   0.16      ...       139657.25318  38.924790 -76.985068   

           PROJECTNOTE                                        PROJECTNAME  \
0                  NaN                                             FedAid   
1                  NaN                                                NaN   
2                  NaN           Fedral Aid Citywide Pavement Restoration   
3  POKA-2003-B-0048-JJ  Citywide Pavement Restoration Contract (Local ...   
4                  NaN                                                NaN   

          PROJECTMGR  PROJECTID        TEAMREP                 EMAIL  \
0    Paul Stephenson       5834        O. Hill                   NaN   
1                NaN       2467            NaN                   NaN   
2  Mohamed Abdullahi       5486   Said Cherifi   said.cherifi@dc.gov   
3    Maduabuchi Udeh       5796  Michael Jelen  Michael.Jelen@dc.gov   
4                NaN       8237            NaN                   NaN   

   CONTACTPHONE  
0  202-671-4591  
1           NaN  
2  202-671-4611  
3  202-671-4542  
4           NaN  

[5 rows x 27 columns]

### 3.1.a  what is its shape and what does that mean?

In [6]:
##The shape of a dataset is a record consisting of the number of observations and the number of attributes in the dataset.

df.shape

(465, 27)

## Homework 3.2
 What are the number of rows in each 'QUADRANT' ?

In [10]:
df.groupby(['QUADRANT']).count()

STREET  FROMINTERSECTION  TOINTERSECTION  NOOFBLOCKS  \
QUADRANT                                                         
BN            12                12              12          12   
NE           163               163             163         163   
NW           195               195             195         195   
SE            84                84              84          84   
SW            11                11              11          11   

          WORKDESCRIPTION  YEARBUDGETED  PERCENTCOMPLETED  STATUS  MILES  \
QUADRANT                                                                   
BN                     12            12                12      12     12   
NE                    163           163               163     163    163   
NW                    195           195               195     195    195   
SE                     84            84                84      84     84   
SW                     11            11                11      11     11   

          WARD      ...       YCOORD  LATITUDE  LONGITUDE  PROJECTNOTE  \
QUADRANT            ...                                                  
BN          12      ...           12        12         12            4   
NE         163      ...          163       163        163            8   
NW         193      ...          195       195        195           77   
SE          84      ...           84        84         84            8   
SW          11      ...           11        11         11            2   

          PROJECTNAME  PROJECTMGR  PROJECTID  TEAMREP  EMAIL  CONTACTPHONE  
QUADRANT                                                                    
BN                  8           8         12        8      8             8  
NE                 65          49        163       42     34            43  
NW                147         141        195      135     79           132  
SE                 40          30         84       21      3            22  
SW                  5           6         11        5      4             4  

[5 rows x 26 columns]

BN: 12
NE: 163
NW: 195
SE: 84
SW: 11

## Homework 3.3 - Array math demonstration 
For two arrarys  

a= [1,2,3,4]  type=float

b= [5,6,7,8]  type=float

Peform the following array operations using numpy   
( show both operational use of numpy and functional (example addition operation => + vs addition function => numbpy.add() )

### 3.3.1 addition a+b 

In [26]:
import numpy as np

a = np.array([1.0,2.0,3.0,4.0])
b = np.array([5.0,6.0,7.0,8.0])

print(a + b)
print(np.add(a,b))




[  6.   8.  10.  12.]
[  6.   8.  10.  12.]


### 3.3.2 subtraction a-b

In [27]:
print(a-b)
print(np.subtract(a,b))

[-4. -4. -4. -4.]
[-4. -4. -4. -4.]


### 3.3.3 multiplication a*b

In [28]:
print(a*b)
print(np.multiply(a,b))

[  5.  12.  21.  32.]
[  5.  12.  21.  32.]


### 3.3.4 divsion a/b

In [29]:
print(a/b)
print(np.divide(a, b))

[ 0.2         0.33333333  0.42857143  0.5       ]
[ 0.2         0.33333333  0.42857143  0.5       ]


### 3.3.5 modulo  a%b

In [30]:
print(a%b)
print(np.mod(a,b))

[ 1.  2.  3.  4.]
[ 1.  2.  3.  4.]


### 3.3.6 power  a^b

In [31]:
print(a**b)
print(np.power(a,b))

[  1.00000000e+00   6.40000000e+01   2.18700000e+03   6.55360000e+04]
[  1.00000000e+00   6.40000000e+01   2.18700000e+03   6.55360000e+04]


## Homework 3.4
Find your own data and load it into a dataframe

In [70]:
## Here is a data table that has the median starting salaries and mid career salaries of the graduates of
## 270 universities in the USA. The dataset also contains mid-career 10th, 25th, 75th and 90th career salaries 
##The universities are categorized into the following groups: Engineering, Party, 
## Ivey League, and State

df = pd.read_csv("salaries-by-college-type.csv")
df= df.replace({'\$':''}, regex = True)
df= df.replace({'\,':''}, regex = True)
df["Starting Median Salary"]= df["Starting Median Salary"].astype('float64', ignore = False)
df["Mid-Career Median Salary"]= df["Mid-Career Median Salary"].astype('float64', ignore = False)
df["Mid-Career 90th Percentile Salary"] = df["Mid-Career 90th Percentile Salary"].astype('float64', ignore = False)


df.head()

School Name  School Type  \
0  Massachusetts Institute of Technology (MIT)  Engineering   
1     California Institute of Technology (CIT)  Engineering   
2                          Harvey Mudd College  Engineering   
3  Polytechnic University of New York Brooklyn  Engineering   
4                                 Cooper Union  Engineering   

   Starting Median Salary  Mid-Career Median Salary  \
0                 72200.0                  126000.0   
1                 75500.0                  123000.0   
2                 71800.0                  122000.0   
3                 62400.0                  114000.0   
4                 62200.0                  114000.0   

  Mid-Career 10th Percentile Salary Mid-Career 25th Percentile Salary  \
0                          76800.00                          99200.00   
1                               NaN                         104000.00   
2                               NaN                          96000.00   
3                          66800.00                          94300.00   
4                               NaN                          80200.00   

  Mid-Career 75th Percentile Salary  Mid-Career 90th Percentile Salary  
0                         168000.00                           220000.0  
1                         161000.00                                NaN  
2                         180000.00                                NaN  
3                         143000.00                           190000.0  
4                         142000.00                                NaN

## Homework 3.5
Provide an interesting analysis of the data columns ( frequency or averages )

In [93]:
## Here are the the average median starting salraries, average median mid-career salaries, 
## and average 90th percentile mid-career salaries across all the universities

df.mean()

Starting Median Salary                46068.401487
Mid-Career Median Salary              83932.342007
Mid-Career 90th Percentile Salary    157705.627706
dtype: float64

In [94]:
##Which universities had the greatest and lowest median starting salaries:
print("University with greatest median starting salary: ", df.loc[df['Starting Median Salary'].idxmax()][0],
       " ", df["Starting Median Salary"].max())
print("University with lowest median starting salary: ", df.loc[df['Starting Median Salary'].idxmin()][0],
       " ", df["Starting Median Salary"].min())

University with greatest median starting salary:  California Institute of Technology (CIT)   75500.0
University with lowest median starting salary:  Morehead State University   34800.0


In [95]:
##Which universities had the greatest and lowest median mid-career salaries:
print("University with greatest median mid-career salary: ", df.loc[df['Mid-Career Median Salary'].idxmax()][0],
       " ", df["Mid-Career Median Salary"].max())
print("University with lowest median mid-career salary: ", df.loc[df['Mid-Career Median Salary'].idxmin()][0],
       " ", df["Mid-Career Median Salary"].min())

University with greatest median mid-career salary:  Dartmouth College   134000.0
University with lowest median mid-career salary:  Black Hills State University   43900.0


In [100]:
print((134000/75500 - 1)*100)
print((43900/34800 - 1)*100)

##It's intersting that the greatest mid-career median salary is %77 higher than the greatest starting median salary,
##But the lowest mid-career median salary is only %26 higher than the lowest starting median salary

77.48344370860927
26.14942528735633


In [98]:
##We can also compare how the universities differ by type:
##Interesting that hte Mid-Career 90th Percentile salary for Liberal Arts colleges is higher than it is for Engineering colleges
##But the Engineering colleges have a higher Mid-Career median salary.

df.groupby(['School Type']).mean()

Starting Median Salary  Mid-Career Median Salary  \
School Type                                                      
Engineering             59057.894737             103842.105263   
Ivy League              60475.000000             120125.000000   
Liberal Arts            45746.808511              89378.723404   
Party                   45715.000000              84685.000000   
State                   44126.285714              78567.428571   

              Mid-Career 90th Percentile Salary  
School Type                                      
Engineering                       173333.333333  
Ivy League                        269625.000000  
Liberal Arts                      191142.857143  
Party                             166947.368421  
State                             147571.428571